### ISSUE TITLE
Mastodon v3.3.0 Google Cloud Storage Configuration Not Working

### ISSUE THREAD

**Issue by gsugambit** [mastodon/mastodon#17435]
### Steps to reproduce the problem

1. I have configured the .env.production file to have the proper configuration for google cloud storage
![image](https://user-images.githubusercontent.com/13861767/152403859-35d7b9f3-b713-458d-8e4b-3e3a1eca9d1f.png)
2. After I start the mastodon application (im used docker), if I try to upload a toot with an image, I get a 500 internal server error. I've noticed that my google cloud storage bucket is also empty. Previously this worked for me.
![image](https://user-images.githubusercontent.com/13861767/152404448-ec7cc098-047f-449e-8ebb-bb92b32a3423.png)
3. The weird issue is in the past this used to work. I believe i changed the keys because i switched to a different google account and nothing has worked since. I have confirmed the keyID and secretAccessKey are absolutely correct along with the s3 bucket name.
![image](https://user-images.githubusercontent.com/13861767/152404521-8e1b709c-834d-460a-9662-afeb963b6490.png)

### Expected behaviour

Expect images to show up in browser and mobile application.

### Actual behaviour

Images do not show and it seems like mastodon is not uploading images to the storage bucket.

### Specifications

3.3.0

**Comment by gsugambit** [comment]
Where exactly could i access the logs for this?

**Comment by gsugambit** [comment]
I noticed in the web logs, somewhere an invalid argument exception is being thrown but it's not too descriptive

`eb_1        | [677a086a-6224-41ef-b3e2-63be84cc5f07] method=POST path=/inbox format=*/* controller=ActivityPub::InboxesController action=create status=500 error='Aws::S3::Errors::InvalidArgument: Invalid argument.' duration=4848.44 view=0.00 db=8.92 key=https://node9.org/channel/linus
web_1        | [677a086a-6224-41ef-b3e2-63be84cc5f07]   
web_1        | [677a086a-6224-41ef-b3e2-63be84cc5f07] Aws::S3::Errors::InvalidArgument (Invalid argument.):
web_1        | [677a086a-6224-41ef-b3e2-63be84cc5f07]   
web_1        | [677a086a-6224-41ef-b3e2-63be84cc5f07] lib/paperclip/attachment_extensions.rb:51:in `block in save'
web_1        | [677a086a-6224-41ef-b3e2-63be84cc5f07] lib/paperclip/attachment_extensions.rb:57:in `save'
web_1        | [677a086a-6224-41ef-b3e2-63be84cc5f07] app/models/account.rb:345:in `save_with_optional_media!'
web_1        | [677a086a-6224-41ef-b3e2-63be84cc5f07] app/services/activitypub/process_account_service.rb:79:in `update_account'
web_1        | [677a086a-6224-41ef-b3e2-63be84cc5f07] app/services/activitypub/process_account_service.rb:28:in `block in call'
web_1        | [677a086a-6224-41ef-b3e2-63be84cc5f07] app/services/activitypub/process_account_service.rb:19:in `call'
web_1        | [677a086a-6224-41ef-b3e2-63be84cc5f07] app/services/activitypub/fetch_remote_account_service.rb:31:in `call'
web_1        | [677a086a-6224-41ef-b3e2-63be84cc5f07] app/services/activitypub/fetch_remote_key_service.rb:37:in `find_account'
web_1        | [677a086a-6224-41ef-b3e2-63be84cc5f07] app/services/activitypub/fetch_remote_key_service.rb:24:in `call'
web_1        | [677a086a-6224-41ef-b3e2-63be84cc5f07] app/controllers/concerns/signature_verification.rb:221:in `block in account_from_key_id'
web_1        | [677a086a-6224-41ef-b3e2-63be84cc5f07] app/controllers/concerns/signature_verification.rb:234:in `stoplight_wrap_request'
web_1        | [677a086a-6224-41ef-b3e2-63be84cc5f07] app/controllers/concerns/signature_verification.rb:221:in `account_from_key_id'
web_1        | [677a086a-6224-41ef-b3e2-63be84cc5f07] app/controllers/concerns/signature_verification.rb:81:in `signed_request_account'
web_1        | [677a086a-6224-41ef-b3e2-63be84cc5f07] app/controllers/concerns/signature_verification.rb:49:in `require_signature!'
`

**Comment by ClearlyClaire** [comment]
First, update to 3.3.2, there are important security fixes. There are also S3-related fixes but I doubt they apply to your case.

If you still experience issues after upgrading to 3.3.2, try editing `config/initializers/backtrace_silencers.rb` to uncomment `Rails.backtrace_cleaner.remove_silencers!`, it might give more information in upcoming logs.

**Comment by arttuvann** [comment]
I am having this same issue with v3.5.0 running directly in a VM, connecting to Google Cloud Storage. To my understanding I've configured the bucket and .env.production accordingly.

journalctl -u mastodon-web

```
Apr 06 18:10:02 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] [paperclip] Trying to link /tmp/RackMultipart20220406-32241-bggvgb.gif to /tmp/32f785a6526ae12956b257424d19744e20220406-32241-mjxf2f.gif
Apr 06 18:10:02 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] [paperclip] Trying to link /tmp/32f785a6526ae12956b257424d19744e20220406-32241-mjxf2f.gif to /tmp/75944864f903fe0e1fcb6acf451b1a5020220406-32241-z6u0$
Apr 06 18:10:02 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] Command :: file -b --mime '/tmp/75944864f903fe0e1fcb6acf451b1a5020220406-32241-z6u08g.gif'
Apr 06 18:10:02 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] Command :: ffmpeg -nostdin -ss '0' -i '/tmp/32f785a6526ae12956b257424d19744e20220406-32241-mjxf2f.gif' -loglevel 'fatal' -vf 'scale='\''min(400\, iw)$
Apr 06 18:10:02 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] Command :: convert '/tmp/32f785a6526ae12956b257424d19744e20220406-32241-mjxf2f20220406-32241-2jjut1.png' RGB:-
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] [paperclip] Trying to link /tmp/32f785a6526ae12956b257424d19744e20220406-32241-mjxf2f20220406-32241-2jjut1.png to /tmp/42f1194ad50373047235f7783a197f$
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] [paperclip] saving media_attachments/files/108/086/467/178/129/560/original/182a074b644f7e1f.gif
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] method=POST path=/api/v2/media format=html controller=Api::V2::MediaController action=create status=500 error='Aws::S3::Errors::InvalidArgument: Inva$
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e]
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] Aws::S3::Errors::InvalidArgument (Invalid argument.):
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e]
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] aws-sdk-core (3.127.0) lib/seahorse/client/plugins/raise_response_errors.rb:17:in `call'
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] aws-sdk-s3 (1.113.0) lib/aws-sdk-s3/plugins/sse_cpk.rb:24:in `call'
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] aws-sdk-s3 (1.113.0) lib/aws-sdk-s3/plugins/dualstack.rb:27:in `call'
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] aws-sdk-s3 (1.113.0) lib/aws-sdk-s3/plugins/accelerate.rb:56:in `call'
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] aws-sdk-core (3.127.0) lib/aws-sdk-core/plugins/checksum_algorithm.rb:111:in `call'
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] aws-sdk-core (3.127.0) lib/aws-sdk-core/plugins/jsonvalue_converter.rb:22:in `call'
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] aws-sdk-core (3.127.0) lib/aws-sdk-core/plugins/idempotency_token.rb:19:in `call'
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] aws-sdk-core (3.127.0) lib/aws-sdk-core/plugins/param_converter.rb:26:in `call'
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] aws-sdk-core (3.127.0) lib/seahorse/client/plugins/request_callback.rb:71:in `call'
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] aws-sdk-core (3.127.0) lib/aws-sdk-core/plugins/response_paging.rb:12:in `call'
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] aws-sdk-core (3.127.0) lib/seahorse/client/plugins/response_target.rb:24:in `call'
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] aws-sdk-core (3.127.0) lib/seahorse/client/request.rb:72:in `send_request'
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] aws-sdk-s3 (1.113.0) lib/aws-sdk-s3/client.rb:12367:in `put_object'
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] aws-sdk-s3 (1.113.0) lib/aws-sdk-s3/file_uploader.rb:64:in `block in put_object'
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] aws-sdk-s3 (1.113.0) lib/aws-sdk-s3/file_uploader.rb:53:in `block in open_file'
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] aws-sdk-s3 (1.113.0) lib/aws-sdk-s3/file_uploader.rb:53:in `open'
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] aws-sdk-s3 (1.113.0) lib/aws-sdk-s3/file_uploader.rb:53:in `open_file'
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] aws-sdk-s3 (1.113.0) lib/aws-sdk-s3/file_uploader.rb:63:in `put_object'
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] aws-sdk-s3 (1.113.0) lib/aws-sdk-s3/file_uploader.rb:45:in `upload'
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] aws-sdk-s3 (1.113.0) lib/aws-sdk-s3/customizations/object.rb:440:in `upload_file'
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] kt-paperclip (7.1.1) lib/paperclip/storage/s3.rb:383:in `block in flush_writes'
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] kt-paperclip (7.1.1) lib/paperclip/storage/s3.rb:356:in `each'
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] kt-paperclip (7.1.1) lib/paperclip/storage/s3.rb:356:in `flush_writes'
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] kt-paperclip (7.1.1) lib/paperclip/attachment.rb:240:in `save'
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] lib/paperclip/attachment_extensions.rb:84:in `block in save'
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] stoplight (2.2.1) lib/stoplight/light/runnable.rb:51:in `run_code'
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] stoplight (2.2.1) lib/stoplight/light/runnable.rb:35:in `run_green'
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] stoplight (2.2.1) lib/stoplight/light/runnable.rb:23:in `run'
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] lib/paperclip/attachment_extensions.rb:90:in `save'
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] kt-paperclip (7.1.1) lib/paperclip/has_attached_file.rb:92:in `block in add_active_record_callbacks'
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] activesupport (6.1.5) lib/active_support/callbacks.rb:427:in `instance_exec'
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] activesupport (6.1.5) lib/active_support/callbacks.rb:427:in `block in make_lambda'
Apr 06 18:10:03 mastodon-server bundle[32185]: [8db90db4-c1fa-4f74-b441-1fcccacd2f7e] activesupport (6.1.5) lib/active_support/callbacks.rb:235:in `block in halting_and_conditional'
```

**Comment by arttuvann** [comment]
In the end it was a config problem on the cloud side. With v3.5.1 and object-level ACLs enabled, the uploading works.

**Comment by jaredallard** [comment]
@arttuvann I just want you to know I literally love you. Sure, it took me literal hours of googling to figure this out, but holy crap thank you.

**Comment by evelritual** [comment]
@arttuvann this also was the resolution to my issue, much thanks!

**Comment by puyoxyz** [comment]
> In the end my case was a config problem on the cloud side. With v3.5.1 and object-level ACLs enabled, the uploading works.

I also had this problem and this was the solution.

For anyone who can't find it: 

Cloud Storage -> Browser -> your mastodon bucket -> Permissions -> Access control, then pick Fine-grained
![image](https://user-images.githubusercontent.com/15311104/170882739-5db6c480-a80e-490e-829f-05a97858d365.png)


Also, maybe this should be added to the docs somewhere?

**Comment by sventiffe** [comment]
@puyoxyz I'm running into a similar issue (albeit the error message differs, see https://github.com/mastodon/mastodon/issues/17435) and have not been able to resolve it. Which other settings do you use for the Google Cloud Storage bucket if I may ask?

**Comment by puyoxyz** [comment]
> @puyoxyz I'm running into a similar issue (albeit the error message differs, see #17435) and have not been able to resolve it. Which other settings do you use for the Google Cloud Storage bucket if I may ask?

What's the error message? The issue you linked is the one we're currently on, maybe you said the wrong issue number

Anyways, here are my settings:

![image](https://user-images.githubusercontent.com/15311104/200470531-704ae640-7933-4086-a814-b0f36e8e4641.png)

**Comment by sventiffe** [comment]
Yes, I linked the wrong issue, I was meant to point to https://github.com/mastodon/mastodon/issues/18067

In my case, the service account that I created through the Cloud Storage settings did not receive the correct role. I've had to assign the objectAdmin role manually. For debugging, I used the command line to manually upload files using the service account credentials. This helped troubleshooting the issue. Plus the information in this issue about the fine grained settings, so thank you folks!
